trie/triestate/state.go

```diff
@@ -16,13 +16,225 @@
 
 package triestate
 
-import "github.com/ethereum/go-ethereum/common"
+import (
+	"errors"
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/ethereum/go-ethereum/trie/trienode"
+)
+
+// Trie is an Ethereum state trie, can be implemented by Ethereum Merkle Patricia
+// tree or Verkle tree.
+type Trie interface {
+	// Get returns the value for key stored in the trie.
+	Get(key []byte) ([]byte, error)
+
+	// Update associates key with value in the trie.
+	Update(key, value []byte) error
+
+	// Delete removes any existing value for key from the trie.
+	Delete(key []byte) error
+
+	// Commit the trie and returns a set of dirty nodes generated along with
+	// the new root hash.
+	Commit(collectLeaf bool) (common.Hash, *trienode.NodeSet)
+}
+
+// TrieLoader wraps functions to load tries.
+type TrieLoader interface {
+	// OpenTrie opens the main account trie.
+	OpenTrie(root common.Hash) (Trie, error)
+
+	// OpenStorageTrie opens the storage trie of an account.
+	OpenStorageTrie(stateRoot common.Hash, addrHash, root common.Hash) (Trie, error)
+}
 
 // Set represents a collection of mutated states during a state transition.
 // The value refers to the original content of state before the transition
 // is made. Nil means that the state was not present previously.
 type Set struct {
 	Accounts   map[common.Address][]byte                 // Mutated account set, nil means the account was not present
 	Storages   map[common.Address]map[common.Hash][]byte // Mutated storage set, nil means the slot was not present
-	Incomplete map[common.Address]struct{}               // Indicator whether the storage slot is incomplete due to large deletion
+	Incomplete map[common.Address]struct{}               // Indicator whether the storage is incomplete due to large deletion
+	size       common.StorageSize                        // Approximate size of set
+}
+
+// New constructs the state set with provided data.
+func New(accounts map[common.Address][]byte, storages map[common.Address]map[common.Hash][]byte, incomplete map[common.Address]struct{}) *Set {
+	return &Set{
+		Accounts:   accounts,
+		Storages:   storages,
+		Incomplete: incomplete,
+	}
+}
+
+// Size returns the approximate memory size occupied by the set.
+func (s *Set) Size() common.StorageSize {
+	if s.size != 0 {
+		return s.size
+	}
+	for _, account := range s.Accounts {
+		s.size += common.StorageSize(common.HashLength + len(account))
+	}
+	for _, slots := range s.Storages {
+		for _, val := range slots {
+			s.size += common.StorageSize(common.HashLength + len(val))
+		}
+		s.size += common.StorageSize(common.HashLength)
+	}
+	s.size += common.StorageSize(common.HashLength * len(s.Incomplete))
+	return s.size
+}
```

Comment:
All the `common.HashLength` instances should be updated with `common.AddressLength`, right?

Reply:
Actually, almost, only 3 out of 4